### PR TITLE
Disable NODERAWFS tests on Mac

### PR DESCRIPTION
**Pull request by aheejin**
NODERAWFS support added by #6008 seems to be still flaky on Mac. (Not sure about Windows. Out waterfall bots have other problems at the moment so they can't test this)

Also add support for platform-dependent tests.

### CODE REVIEW

**Comment by saschanaz** [comment]
I would say that we add/modify conditional preprocessing on the failing unistd test. I think there already is one...

**Comment by saschanaz** [comment]
https://github.com/kripken/emscripten/blob/941e9de0946077d835a40556048af1de4e4e2abe/tests/unistd/unlink.c#L140-L158

We have a couple of platform guards here, can we fix them instead of just disabling failing tests?

**Comment by aheejin** [comment]
Those `#ifdef __APPLE__` guards don't seem to be working for emscripten. Not sure why there are so many of them in the tests/ directory. Emscripten compiler does not predefine `__APPLE__`.

The reason is, in clang, WebAssembly (for Wasm backend) and Emscripten (for fastcomp) have a separate OS target, so when we give `-target wasm32-unknown-unknown-elf` (for Wasm backend) or `-target asmjs-unknown-emscripten` (for fastcomp) to clang, the host OS will not be the actual OS the code is running but a separate "WebAssembly" or "Emscripten" OS.
[fastcomp: clang-predefined macros](https://github.com/kripken/emscripten-fastcomp-clang/blob/29de7f517cc4f0ed17a4123e73844713da6c3430/lib/Basic/Targets.cpp#L329-L347)
[Wasm backend: clang-predefined macros](https://github.com/llvm-mirror/clang/blob/23a20415d5f95650e2e51200580ca97f2df1061c/lib/Basic/Targets/OSTargets.h#L710-L718)

Rather, emscripten predefines `unix`, `__unix`, and `__unix__`.
[fastcomp: predefined unix macros](https://github.com/kripken/emscripten-fastcomp-clang/blob/29de7f517cc4f0ed17a4123e73844713da6c3430/lib/Basic/Targets.cpp#L343-L346)
[Wasm backend: predefined unix macros](https://github.com/kripken/emscripten/blob/941e9de0946077d835a40556048af1de4e4e2abe/tools/shared.py#L962-L965)
#4718 and #4727 argue they add the unix macros because emscripten behaves _similarly_ to unix, but https://github.com/kripken/emscripten/pull/4718#issuecomment-260185896 also has concerns on if it is strictly accurate.

So, in short, OS-related macros like `__APPLE__` do not currently work in emscripten. Do you think we should re-evaluate our macro-defining choices, so we should define macros based on the actual OS the code is running? Or should we just disable the relevant tests as I did in this PR?

Maybe good to cc people in those old issues as well: @kripken @juj @jgravelle-google @dschuff

**Comment by saschanaz** [comment]
Hmm, we may have to accept multiple errors numbers as expected without conditions.

```
assert(errno == EISDIR || errno == EPERM); // Linux throws EISDIR, OS X throws EPERM (for example)
```

**Comment by aheejin** [comment]
There are many conditional guards like `#ifdef __APPLE__` based on the OS-predefined symbols in the emscripten test code. I wonder what they tried to target in the first place then. Do you think we should remove them?

**Comment by dschuff** [comment]
In principle they probably shouldn't be there, and in practice they don't do anything, so it seems like they should probably be removed, (depending on what they are there for, e.g. if they came from third-party code or something).

As for this PR specifically, do we want to duplicate or replace the OS detection logic from shared.py? I get that we need `no_windows()` etc to fit in with the skipping logic but maybe we should leave the `if WINDOWS` bits in this file as they are, and (if we want to replace that) replace it globally in a separate PR.

**Comment by aheejin** [comment]
The `WINDOWS` thing from shared.py is not completely replaced; it is used elsewhere too. I made `is_mac()` and `is_windows()` to allow us to use decorators like `@no_mac` and `@no_windows`, and thoght it would make more sense to consistently use them in this test_core.py.

We have three options:
1. Add `@no_mac` / `@no_windows` / `is_mac()` / `is_windows()`, and use them wherever it is necessary within test_core.py. (The current state) `WINDOWS` from shared.py is also used elsewhere, so don't touch that.
2. Add `@no_mac` / `@no_windows` / `is_mac()` / `is_windows()`, but only use them as decorators like `@no_mac`, and when we want to test the condition in an if statement, we stick to `WINDOWS` or `OSX` defined in shared.py.
3. Don't add anything, and disable tests only with `WINDOWS` or `OSX`. That means even when we want to disable a test as a whole, we need to insert `if`s within the function to use `WINDOWS`/`OSX`. 

If we go with the option 1 as I am currently doing, I think separating it with two PRs might make more sense too.
1. Add @no_mac` / `@no_windows` / `is_mac()` / `is_windows()`, and replace uses within this file to use them
2. Disable NODERAWFS-related tests on Mac
What do you think?

**Comment by kripken** [comment]
Is the issue here just the error codes? If so, can we do what @saschanaz suggested above,
```
assert(errno == EISDIR || errno == EPERM); // Linux throws EISDIR, macOS throws EPERM (for example)
```
or is the OS-specific behavior more complicated to work around in the test?

**Comment by aheejin** [comment]
Oh no, I can do that. It is just a way of disabling a part of the test - I can either check if it's mac or not in test_core.py using `if not is_mac() ...`, or just modify the test c file itself. First I tried to use `__APPLE__` macro but found out it does not work.

**Comment by aheejin** [comment]
@kripken Oh, on second thought, if we are already disabling windows by this code, wouldn't disabling mac in the same manner, within this `if` statement, be more consistent?
https://github.com/kripken/emscripten/blob/e24661497247ecac78166b02e25e880e0ee14c9d/tests/test_core.py#L4711-L4716

**Comment by aheejin** [comment]
Split decorator part in #6128, which should land first.

**Comment by kripken** [comment]
Yeah, if we already have an if for Windows, we can add Mac to there too.

**Comment by saschanaz** [comment]
BTW, can we file an issue on Nodejs (or possibly libuv) side?

```
======================================================================
ERROR: test_fs_writeFile (test_core.binaryen2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/b/build/slave/mac/build/src/src/work/wasm-install/emscripten/tests/test_core.py", line 57, in decorated
    func(self, js_engines=[NODE_JS])
  File "/b/build/slave/mac/build/src/src/work/wasm-install/emscripten/tests/test_core.py", line 4551, in test_fs_writeFile
    self.do_run_from_file(src, out, js_engines=js_engines)
  File "/b/build/slave/mac/build/src/src/work/wasm-install/emscripten/tests/runner.py", line 649, in do_run_from_file
    self.do_run(open(src).read(), open(expected_output).read(), *args, **kwargs)
  File "/b/build/slave/mac/build/src/src/work/wasm-install/emscripten/tests/runner.py", line 689, in do_run
    raise e
Exception: Expected to find 'read a=1
read b=2
read c=3
' in 'read c=3
read b=2
read 
', diff:
--- expected
+++ actual
@@ -1,4 +1,4 @@
-read a=1
+read c=3
 read b=2
-read c=3
+read
```

It seems `fs.open(src, 'a')` doesn't set stream position correctly on Node.js side, I think it's worth filing an issue then.

**Comment by aheejin** [comment]
@saschanaz Are you referring to me in regards to the Nodejs issue?

**Comment by saschanaz** [comment]
Yes, but feel free to ignore if you don't have free time or you only have bot access for macOS tests...

**Comment by aheejin** [comment]
Sorry, I don't have much background on Node, so I'm not sure if I can provide sufficient context there.

**Comment by saschanaz** [comment]
Oops, never mind then, sorry!

**Comment by curiousdannii** [comment]
@saschanaz That's documented at https://nodejs.org/api/fs.html#fs_fs_open_path_flags_mode_callback:

>On Linux, positional writes don't work when the file is opened in append mode. The kernel ignores the position argument and always appends the data to the end of the file.

You need to create the file if it doesn't exist, open as 'r+', and then manually seek to the end.

**Comment by saschanaz** [comment]
IMO that 'positional writes' means giving explicit position argument, whereas the writrFile test expects the append mode automatically sets the position to the EOF.

File: tests/test_core.py [hunk 1]
```diff
@@ -4564,6 +4564,7 @@ def test_fs_trackingdelegate(self):
     out = path_from_root('tests', 'fs', 'test_trackingdelegate.out')
     self.do_run_from_file(src, out)
 
+  @no_mac('NODERAWFS support on mac is flaky')
```
Comment (kripken) [inline tests/test_core.py]: this also disables the non-NODERAWFS part on mac. i guess there isn't a simple way to disable just that part, though, unless we split up this test. not sure it's important enough to justify that.

Comment (aheejin) [reply tests/test_core.py]: I split the test into two.

File: tests/test_core.py [hunk 2]
```diff
@@ -4564,13 +4564,23 @@ def test_fs_trackingdelegate(self):
...
     self.emcc_args += ['-s', 'DISABLE_EXCEPTION_CATCHING=1'] # see issue 2334
     src = path_from_root('tests', 'fs', 'test_writeFile.cc')
     out = path_from_root('tests', 'fs', 'test_writeFile.out')
     self.do_run_from_file(src, out, js_engines=js_engines)
 
+  # NODERAWFS support for OSX is flaky, but it is not desirable to always
+  # disable test_fs_writeFile test, so we split the test into two.
+  # See issue 6121
+  @no_osx('NODERAWFS support on OSX is flaky')
+  @also_with_noderawfs
```
Comment (kripken) [inline tests/test_core.py]: I think that on non-mac this will run the test twice without NODERAWFS and once with it? (since `test_fs_writeFile_noderawfs` is marked as `also_with_noderawfs` which will run it once normally and once with NODERAWFS)

Seems a little wasteful, but there isn't a good and obvious way to fix it. Best I can think of is: go back to what we had before, and in the test check if `mac and is_noderawfs`, and if so skip that test (I think we need to add a method to check `is_noderawfs`).

Comment (aheejin) [reply tests/test_core.py]: Not sure if I can check that, because I guess the way to implement `is_noderawfs` would be to check if `Settings.NODERAWFS == 1`. But that is true only in the child process this test_core.py launches with option `-s NODERAWFS=1` and not in test_core.py.

Comment (kripken) [reply tests/test_core.py]: Hmm, yes, I think we would need to check something like  `'NODERAWFS=1' in self.emcc_args`.

Comment (aheejin) [reply tests/test_core.py]: Done

File: tests/test_core.py [hunk 3]
```diff
@@ -4566,6 +4566,9 @@ def test_fs_trackingdelegate(self):
 
   @also_with_noderawfs
   def test_fs_writeFile(self, js_engines=None):
+    if self.is_osx() and 'NODERAWFS=1' in self.emcc_args:
```
Comment (saschanaz) [inline tests/test_core.py]: Too bad to miss reviewing #6128, but can we rename `is_osx` to `is_macos`, as Apple renamed OS X to macOS to keep consistency with iOS, watchOS and tvOS?

Comment (aheejin) [reply tests/test_core.py]: Done in #6138.

**Comment by saschanaz** [comment]
Now the only thing left is the unlink test. Can you check what errno the test emits on line 90 of tests/unistd/unlink.c, and if possible allow it like this?

```
assert(errno == EISDIR || errno == EPERM); // Linux throws EISDIR, macOS throws EPERM (for example)
```

**Comment by aheejin** [comment]
Even if we do that, the assertion on the line 157 fails too; the errno in this case is `EISDIR`. 
https://github.com/kripken/emscripten/blob/16df331292e156d5578c053c26bc73943904d6d0/tests/unistd/unlink.c#L157

It would look kind of weird if we change that code into 
```
#ifdef __APPLE__
  assert(errno == EISDIR);
#else
  assert(errno == EBUSY || EISDIR);
```
when there is no case the `ifdef __APPLE__` part can be satisfied.

How about wrapping up this PR here and deleting all `__APPLE__` related code in another PR?

**Comment by saschanaz** [comment]
Maybe someone wanted to check native gcc/clang behavior, but :+1: for removing `__APPLE__` codes *and* allowing all OS-dependent behaviors.

**Comment by aheejin** [comment]
Because #6142 has been merged, this should be unnecessary. I'll watch the waterfall to go back to green and then abandon this if everything's OK.

**Comment by aheejin** [comment]
After #6140 and #6142 are merged, this should become unnecessary. I'll check the waterfall until it goes back to green and then abandon this if everything's OK.
